### ISSUE TITLE
Space needed before number to recognize - as operator

### ISSUE THREAD

**Issue by aslakr** [camoy/amath#11]
``` sh
$ echo "a-2b" | build/amath 
<math><mi>a</mi><mn>-2</mn><mi>b</mi></math>
```

``` sh
$ echo "a - 2b" | build/amath
<math><mi>a</mi><mo>&minus;</mo><mn>2</mn><mi>b</mi></math>
```

``` sh
$ echo "a-b" | build/amath
<math><mi>a</mi><mo>&minus;</mo><mi>b</mi></math>
```

I suppose even `-2` should be  `<math><mo>&minus;</mo><mn>2</mn></math>` and not `<math><mn>-2</mn></math>`?

**Comment by camoy** [comment]
This is actually the desired behavior according to the [specification](http://www1.chapman.edu/~jipsen/mathml/asciimathsyntax.html). See the section on numbers. Otherwise it would be impossible to display negative numbers in an appropriate way.

**Comment by aslakr** [comment]
I was just asking because MathJax  [sample-ascimath.html](http://cdn.mathjax.org/mathjax/latest/test/sample-asciimath.html) which contains:

``` html
<p>
When `a != 0`, there are two solutions to `ax^2 + bx + c = 0` and they are
</p>
<p style="text-align:center">
`x = (-b +- sqrt(b^2-4ac))/(2a) .`
</p>
```

ie. `x = (-b +- sqrt(b^2-4ac))/(2a) .`  generates:

``` xml
<math xmlns="http://www.w3.org/1998/Math/MathML">
  <mstyle displaystyle="true">
    <mi>x</mi>
    <mo>=</mo>
    <mfrac>
      <mrow>
        <mo>-</mo>
        <mi>b</mi>
        <mo>&#xB1;</mo>
        <msqrt>
          <mrow>
            <msup>
              <mi>b</mi>
              <mn>2</mn>
            </msup>
            <mo>-</mo>
            <mn>4</mn>
            <mi>a</mi>
            <mi>c</mi>
          </mrow>
        </msqrt>
      </mrow>
      <mrow>
        <mn>2</mn>
        <mi>a</mi>
      </mrow>
    </mfrac>
    <mo>.</mo>
  </mstyle>
</math>
```

while amath

``` xml
<math><mi>x</mi><mo>=</mo><mfrac><mrow><mo>&minus;</mo><mi>b</mi>
<mo>&pm;</mo><msqrt><mrow><msup><mi>b</mi><mn>2</mn></msup>
<mn>-4</mn><mi>a</mi><mi>c</mi></mrow></msqrt></mrow><mrow>
<mn>2</mn><mi>a</mi></mrow></mfrac><mo>.</mo></math>
```

(yes, the period should not have been there, but since it exist in in the original sample, I let it stay)

since MathJax/AsciiMathML.js seems to be the current reference implementation one should probably bug report this to them or if they won't change this, write why amath is implemented as it is. 

Especially since some software are trying amath as an alternative mailmate/asciimath.mmbundle@365c44b1e99523d5ef06691d9b7726abd49258c0 and their users might ask why some things works different from AsciiMathML.js.

**Comment by aslakr** [comment]
Maybe related asciimath/asciimathml#30

**Comment by camoy** [comment]
I see what you're saying. Looking at the MathML spec the recommended style for negative numbers is `<mrow><mo>-</mo><mn>1</mn></mrow>` anyway. Aside from that, having `b^2-4ac` treat the negative as a unary minus is a little awkward. I think perhaps the AsciiMathML.js behavior is better even if it doesn't conform exactly to the original specification. I'll work to change this.

**Comment by camoy** [comment]
The central problem: how can we infer when a minus sign should be treated as unary or binary operator? We could always treat it as binary (which I think is the current AsciiMathML.js approach which is potentially a bug), but that would lead to significant spaces in negative numbers. If we aren't determining this using whitespace, then we'll have to come up with a different way.
- We want it to be treated as unary in `-1`, `(-1,-1)`, `x<-1`, `sqrt(-1)`, `((-1,-1),(-1,-1))`.
- We want it to be treated as binary in `x-1`, `sqrt(1)-1`, `1/2-1`.

It seems that we could potentially distinguish these by specifying that if `"-" NUMBER` occurs at the beginning of an expression or before an operator, they should be wrapped in `mrow`.

Another question: should the `-` in `-a` also be treated as unary? What about other expressions that aren't identifiers or numbers? I'm inclined to say no, but I'm actually not sure.

What do you think?

**Comment by aslakr** [comment]
I’m not sure either.  

I have tried to skim through the MathML spec. First there is something called [Inferred `<mrow>`s](http://www.w3.org/TR/MathML3/chapter3.html#presm.inferredmrow), but lets pretend that doesn't exist.

I'm guess suggestion is that in theory those examples may/should be:

`-1`  =

``` xml
<mrow><mo>&minus;</mo><mn>1</mn></mrow>
```

`(-1,-1)` =

``` xml
<mrow>
  <mo> ( </mo>
  <mrow>
    <mrow>
      <mo> &minus; </mo>
      <mn> 1 </mn>
    </mrow>
    <mo> , </mo>
    <mrow>
      <mo> &minus; </mo>
      <mn> 1 </mn>
    </mrow>
  <mrow>
  <mo> ) </mo>
</mrow>
```

`x<-1` =

``` xml
<mrow>
  <mi> x </mi>
  <mo> &lt; </mo>
  <mrow>
    <mo> &minus; </mo>
    <mn> 1 </mn>
  </mrow>
</mrow>
```

`sqrt(-1)` = 

``` xml
<msqrt>
  <mrow>
    <mrow>
      <mo> &minus; </mo>
      <mn> 1 </mn>
    </mrow>
  </mrow>
</msqrt>
```

(that looks a bit strange, since `<mrow>` normally is inferred and probably is wrong).

maybe it looks better if the example is `sqrt(-1-1)` =

``` xml
<msqrt>
  <mrow>
    <mrow>
      <mo> &minus; </mo>
      <mn> 1 </mn>
    </mrow>
    <mo> &minus; </mo>
    <mn> 1 </mn>
  </mrow>
</msqrt>
```

?

I guess `-a` = `<mrow><mo>-</mo><mi>a</mi></mrow>`

**Comment by aslakr** [comment]
And then there is [Invisible operators](http://www.w3.org/TR/MathML3/chapter3.html#presm.invisibleops):  `&ApplyFunction;`/`&af;`, `&InvisibleTimes;`/`&it;`, `&InvisibleComma;`/`&ic;` and invisible plus `&#x2064;`/`&#8292;`.

**Comment by aslakr** [comment]
Just visualize the difference:

``` html
<!DOCTYPE html>
<html>
<head>
<title>unary/binary and mrow</title>
<style type="text/css">
  body {font-size: xx-large;}
  mrow {background-color: rgba(255, 0, 0, 0.3);}
</style>
</head>
<body>
<p><kbd>(-1,-1)</kbd></p>

<p>
<math>
    <mo>(</mo>
    <mrow>
        <mrow>
            <mo>&minus;</mo>
            <mn>1</mn>
        </mrow>
        <mo>,</mo>
        <mrow>
            <mo>&minus;</mo>
            <mn>1</mn>
        </mrow>
    </mrow>
    <mo>)</mo>
</math>
</p>

<p>
<math>
    <mo>(</mo>
    <mrow>
        <mo>&minus;</mo>
        <mn>1</mn>
        <mo>,</mo>
        <mo>&minus;</mo>
        <mn>1</mn>
    </mrow>
    <mo>)</mo>
</math>
</p>

<p>
<math>
    <mrow>
        <mo>(</mo>
        <mo>&minus;</mo>
        <mn>1</mn>
        <mo>,</mo>
        <mo>&minus;</mo>
        <mn>1</mn>
        <mo>)</mo>
    </mrow>
</math>
</p>

<p><kbd>x&lt;-1</kbd></p>

<p>
<math>
    <mi>x</mi>
    <mo>&lt;</mo>
    <mo>&minus;</mo>
    <mn>1</mn>
</math>
</p>

<p>
<math>
    <mi>x</mi>
    <mo>&lt;</mo>
    <mrow>
        <mo>&minus;</mo>
        <mn>1</mn>
    </mrow>
</math>
</p>
</body>
</html>
```

which gives the following result in Firefox (the difference isn’t that apparent currently in WebKit):

![mrow](https://cloud.githubusercontent.com/assets/80737/14576401/0b03724e-0369-11e6-9b99-1bbd8ad3bcdf.png)

**Comment by aslakr** [comment]
How about `-1^2`?

`-1^2 = {:-1^2:} = -1 != {:-1:}^2 = ({:-1:})^2 = 1`

≈

``` xml
<math>
    <mo>&minus;</mo>
    <msup>
        <mn>1</mn>
        <mn>2</mn>
    </msup>
    <mo>=</mo>
    <mrow>
        <mo>&minus;</mo>
        <msup>
            <mn>1</mn>
            <mn>2</mn>
        </msup>
    </mrow>
    <mo>=</mo>
    <mrow>
        <mo>&minus;</mo>
        <mn>1</mn>
    </mrow>
    <mo>&NotEqual;</mo>
    <msup>
        <mrow>
            <mo>&minus;</mo>
            <mn>1</mn>
        </mrow>
        <mn>2</mn>
    </msup>
    <mo>=</mo>
    <msup>
        <mrow>
            <mo>(</mo>
            <mrow>
                <mo>&minus;</mo>
                <mn>1</mn>
            </mrow>
            <mo>)</mo>
        </mrow>
        <mn>2</mn>
    </msup>
    <mo>=</mo>
    <mn>1</mn>
</math>
```

?

**Comment by camoy** [comment]
I've attempted a more intelligent detection of negative signs on the [negative branch](https://github.com/camoy/amath/tree/negative). How does it look to you?

**Comment by drlippman** [comment]
FWIW, AsciiMathML doesn't distinguish between `a - b` and `a-b` anymore.  I don't think is has for a really long time (maybe 2004, with one of the early revisions).  Peter wasn't great about remembering to update his documentation :)

**Comment by camoy** [comment]
Thanks for the clarification. Ignoring the whitespace between operator and operand is probably the right way to go.

I've merged the negative branch which treats whitespace as AsciiMathML.js does. They are not semantically equivalent because amath will generate some explicit `<mrow>` tags that AsciiMathML.js may not, but they should be mostly equivalent visually (except I have resolved the tuple spacing problems raised in asciimath/asciimathml#30).

I believe this solution works reasonably well, but if there are situations where it is incorrect we can resolve it in another issue.
